File: backend/internal/fracindex/fracindex_test.go
package fracindex

import "testing"

func TestGenerateKeyBetween(t *testing.T) {
	cases := []struct {
		name  string
		left  string
		right string
	}{
		{name: "default", left: "", right: ""},
		{name: "before first", left: "", right: "500000000000"},
		{name: "after last", left: "500000000000", right: ""},
		{name: "between adjacent", left: "500000000000", right: "500000000001"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := GenerateKeyBetween(tc.left, tc.right)
			if err != nil {
				t.Fatalf("GenerateKeyBetween returned error: %v", err)
			}
			if tc.left != "" && !(got > tc.left) {
				t.Fatalf("expected %q to be greater than %q", got, tc.left)
			}
			if tc.right != "" && !(got < tc.right) {
				t.Fatalf("expected %q to be less than %q", got, tc.right)
			}
		})
	}
}

func TestGenerateKeyBetweenRepeatedInsertions(t *testing.T) {
	left := "500000000000"
	right := "500000000001"

	for range 32 {
		got, err := GenerateKeyBetween(left, right)
		if err != nil {
			t.Fatalf("GenerateKeyBetween returned error: %v", err)
		}
		if !(got > left && got < right) {
			t.Fatalf("expected %q to be between %q and %q", got, left, right)
		}
		right = got
	}
}

File: frontend/src/features/tasks/utils/fracindex.test.ts
import { describe, expect, it } from 'vitest';
import { generateKeyBetween } from './fracindex';

describe('generateKeyBetween', () => {
  const cases: { name: string; left: string; right: string }[] = [
    { name: 'default', left: '', right: '' },
    { name: 'before first', left: '', right: '500000000000' },
    { name: 'after last', left: '500000000000', right: '' },
    { name: 'between adjacent', left: '500000000000', right: '500000000001' },
  ];

  it.each(cases)('$name', ({ left, right }) => {
    const got = generateKeyBetween(left, right);
    if (left !== '') expect(got > left).toBe(true);
    if (right !== '') expect(got < right).toBe(true);
  });

  it('handles repeated insertions between adjacent keys', () => {
    const left = '500000000000';
    let right = '500000000001';

    for (let i = 0; i < 32; i++) {
      const got = generateKeyBetween(left, right);
      expect(got > left && got < right).toBe(true);
      right = got;
    }
  });

  it('throws on invalid bounds', () => {
    expect(() => generateKeyBetween('b', 'a')).toThrow('invalid bounds');
    expect(() => generateKeyBetween('a', 'a')).toThrow('invalid bounds');
  });

  it.each([
    { left: '', right: '', want: 'O' },
    { left: '', right: '500000000000', want: '+' },
    { left: '500000000000', right: '', want: '50000000000W' },
    { left: '500000000000', right: '500000000001', want: '500000000000O' },
    { left: '!', right: '"', want: '!O' },
    { left: '', right: '!!', want: '!!O' },
    { left: '~', right: '', want: '~O' },
    { left: 'a', right: 'c', want: 'b' },
    { left: 'a', right: 'b', want: 'aO' },
  ])('matches Go output for ("$left", "$right")', ({ left, right, want }) => {
    expect(generateKeyBetween(left, right)).toBe(want);
  });
});
